src/lua/protobuf.lua

```diff
@@ -0,0 +1,624 @@
+    end
+    if max_key ~= key_count then
+        error(('Input array for %q repeated field has inconsistent keys. ' ..
+            'Got table with %d fields and max index of %d'):format(
+            field_def.name, key_count, max_key))
+    end
+end
+
+local function validate_type(field_def, value, exp_type)
+    if type(exp_type) == 'table' then
+        local found = false
+        for _, exp_t in pairs(exp_type) do
+            if type(value) == exp_t then
+                found = true
+                break
+            end
+        end
+        if not found then
+            error(('Field %q of %q type gets %q type value.'):format(
+                field_def.name, field_def.type, type(value)))
+        end
+        return
+    end
+    assert(type(exp_type) == 'string')
+    if type(value) ~= exp_type then
+        error(('Field %q of %q type gets %q type value.'):format(
+            field_def.name, field_def.type, type(value)))
+    end
+    return
+end
+
+local function validate_value_error(field_def, value)
+    error(('Input data for %q field is %q and do not fit in %q')
+        :format(field_def.name, value, field_def.type))
+end
+
+local function validate_value(field_def, value, range)
+    local min = range ~= nil and range[1] or nil
+    local max = range ~= nil and range[2] or nil
+    if min ~= nil and value < min or max ~= nil and value > max then
+        validate_value_error(field_def, value)
+    end
+end
+
+validate_scalar = function(field_def, value)
+    local scalar_def = scalars[field_def.type]
+    assert(scalar_def.accept_type ~= nil)
+    validate_type(field_def, value, scalar_def.accept_type)
+    if scalar_def.accept_type == 'string' then
+        validate_length(value)
+    end
+    if type(value) == 'number' and is_nan(value) then
+        error(('Input data for %q field is NaN'):format(field_def.name))
+    end
+    if type(value) == 'number' and is_inf(value) then
+        error(('Input data for %q field is inf'):format(field_def.name))
+    end
+    if scalar_def.integral_only then
+        check_integer(field_def, value)
+    end
+    if scalar_def.limits ~= nil then
+        local value_type = type(value)
+        if value_type == 'cdata' then
+            value_type = ffi.istype(int64_t, value) and 'int64'
+                or 'uint64'
+        end
+        validate_value(field_def, value, scalar_def.limits[value_type])
+    end
+end
+
+-- }}} Validation
+
+
+-- {{{ Scalar type definitions
+
+scalars.float = {
+    accept_type = 'number',
+    encode_as_packed = true,
+    integral_only = false,
+    limits = {
+        number = {-3.4028234E+38, 3.4028234E+38},
+    },
+    encode = wireformat.encode_float,
+}
+
+scalars.fixed32 = {
+    accept_type = {'number', 'cdata'},
+    encode_as_packed = true,
+    integral_only = true,
+    limits = {
+        number = {0, 2^32 - 1},
+        int64 = {0LL, 4294967295LL},
+        uint64 = {nil, 4294967295ULL},
+    },
```

Comment:
Is there a reason why `number` values are given in a `2^X` or `2^X - 1` form, while the 64-bit numbers are given as literals? We can do the following:

```lua
limits = {
    number = {0, 2^32 - 1},
    int64 = {0LL, 2LL^32 - 1},
    uint64 = {nil, 2ULL^32 - 1},
},
```

It looks more human readable, as for me.

Reply:
Brought all the limits to the same `2^X` format